src/main/scala/com/cognite/sdk/scala/v1/resources/dataModels.scala

```diff
@@ -60,28 +56,78 @@ class DataModels[F[_]](val requestSession: RequestSession[F])
 }
 
 object DataModels {
-  implicit val dataModelPropertyIndexEncoder: Encoder[DataModelPropertyIndex] =
-    deriveEncoder[DataModelPropertyIndex]
+  implicit val dataModelIdentifierEncoder: Encoder[DataModelIdentifier] =
+    Encoder
+      .encodeIterable[String, Seq]
+      .contramap[DataModelIdentifier](dmi =>
+        dmi match {
+          case DataModelIdentifier(Some(space), model) =>
+            Seq(space, model)
+          case DataModelIdentifier(_, model) =>
+            Seq(model)
+        }
+      )
+
+  implicit val bTreeIndexEncoder: Encoder[BTreeIndex] =
+    deriveEncoder[BTreeIndex]
+  implicit val dataModelPropertyTypeEncoder: Encoder[PropertyType] =
+    Encoder.encodeString.contramap[PropertyType](_.code)
+  implicit val dataModelPropertyIndexesEncoder: Encoder[DataModelIndexes] =
+    deriveEncoder[DataModelIndexes]
   implicit val dataModelPropertyEncoder: Encoder[DataModelProperty] =
     deriveEncoder[DataModelProperty]
-  implicit val dataModelEncoder: Encoder[DataModel] = deriveEncoder[DataModel]
-  implicit val dataModelItemsEncoder: Encoder[Items[DataModel]] = deriveEncoder[Items[DataModel]]
+  implicit val contstrainedPropertyEncoder: Encoder[ContstrainedProperty] =
+    deriveEncoder[ContstrainedProperty]
+  implicit val uniquenessConstraintEncoder: Encoder[UniquenessConstraint] =
+    deriveEncoder[UniquenessConstraint]
+  implicit val dataModelConstraintsEncoder: Encoder[DataModelConstraints] =
+    deriveEncoder[DataModelConstraints]
+  implicit val dataModelEncoder: Encoder[DataModel] =
+    deriveEncoder[DataModelDTO].contramap[DataModel](_.toDTO)
+  implicit val dataModelItemsEncoder: Encoder[SpacedItems[DataModel]] =
+    deriveEncoder[SpacedItems[DataModel]]
+  implicit val cogniteIdSpacedItemsEncoder: Encoder[SpacedItems[CogniteId]] =
+    deriveEncoder[SpacedItems[CogniteId]]
   implicit val dataModelListInputEncoder: Encoder[DataModelListInput] =
     deriveEncoder[DataModelListInput]
 
-  implicit val dataModelPropertyIndexDecoder: Decoder[DataModelPropertyIndex] =
-    deriveDecoder[DataModelPropertyIndex]
+  implicit val dataModelIdentifierDecoder: Decoder[DataModelIdentifier] =
+    Decoder
+      .decodeIterable[String, List]
+      .map(ids =>
+        ids match {
+          case modelExternalId :: Nil =>
+            DataModelIdentifier(modelExternalId)
+          case spaceExternalId :: modelExternalId :: Nil =>
+            DataModelIdentifier(spaceExternalId, modelExternalId)
```

Comment:
So we can get rid of the need for the object from above, with this change:
```suggestion
          case modelExternalId :: Nil =>
            DataModelIdentifier(None, modelExternalId)
          case spaceExternalId :: modelExternalId :: Nil =>
            DataModelIdentifier(Some(spaceExternalId), modelExternalId)
```

Super nice stuff, thanks scala :bow:

Reply:
Also not how the matching on `Some(spaceExternalId)` now gives us `spaceExternalId` as `String` instead of `Some(String)`!

Reply:
@silvavelosa: case classes basically come with the [unapply function](https://docs.scala-lang.org/tour/extractor-objects.html).
In reality case classes use some compiler tricks instead, Extractor objects are just in there to allow us to build the same functionality into other stuff as well.. they always hurt my brain when used other places, but really nice for the match-case unpacking